prepare.xml

```diff
@@ -0,0 +1,116 @@
+        <antcall target="makeToolsDir"/>
+        <antcall target="ideaIU.fetch"/>
+        <antcall target="grammarKit.fetch"/>
+        <antcall target="psi.fetch"/>
+    </target>
+
+    <target name="setup" >
+        <antcall target="ideaIU.extract"/>
+        <antcall target="grammarKit.extract"/>
+        <antcall target="psi.extract"/>
+    </target>
+
+    <target name="cleanup" >
+        <delete dir="${toolsDir}"/>
+    </target>
+
+    <target name="makeToolsDir" depends="toolsDir.check" unless="toolsDir.exists">
+        <mkdir dir="${toolsDir}" />
+    </target>
+
+    <!--DOWNLOAD-->
+
+    <target name="ideaIU.fetch" depends="ideaIU.check" unless="ideaIU.exists">
+        <get dest="${ideaIUFile}"
+            src="http://download.jetbrains.com/idea/ideaIU-${ideaVersion}.tar.gz"/>
+    </target>
+
+    <target name="grammarKit.fetch" depends="grammarKit.check" unless="grammarKit.exists">
+        <get dest="${grammarKitFile}"
+            src="https://github.com/JetBrains/Grammar-Kit/releases/download/${grammarKitVersion}/GrammarKit.zip"/>
+    </target>
+
+    <target name="psi.fetch" depends="psi.check" unless="psi.exists">
+        <get dest="${psiFile}"
+             src="https://github.com/JetBrains/Grammar-Kit/releases/download/${grammarKitVersion}/light-psi-all.jar"/>
+    </target>
+
+    <!--FILE EXISTS CHECKS-->
+
+    <target name="toolsDir.check">
+        <condition property="toolsDir.exists">
+            <available file="${toolsDir}" type="dir"/>
+        </condition>
+    </target>
+
+    <target name="ideaIU.check">
+        <condition property="ideaIU.exists">
+            <available file="${ideaIUFile}" type="file"/>
+        </condition>
+    </target>
+
+    <target name="grammarKit.check">
+        <condition property="grammarKit.exists">
+            <available file="${grammarKitFile}" type="file"/>
+        </condition>
+    </target>
+    <target name="psi.check">
+        <condition property="psi.exists">
+            <available file="${psiFile}" type="file"/>
+        </condition>
+    </target>
+
+
+    <!--EXTRACT FILES-->
+
+    <target name="ideaIU.extract">
+        <untar src="${ideaIUFile}" dest="${toolsDir}/idea-IU" compression="gzip"/>
+        <dirset dir="${toolsDir}/idea-IU/" includes="idea-IU*" id="wildCardDir" />
+        <pathconvert pathsep=", " property="ideaDir" refid="wildCardDir" />
+        <copy todir="${basedir}/idea-IU/">
+            <fileset  dir="${ideaDir}" includes="**/*"/>
+        </copy>
+    </target>
+
+    <target name="grammarKit.extract">
+        <unzip src="${grammarKitFile}" dest="${toolsDir}"/>
+        <copy file="${toolsDir}/GrammarKit/lib/grammar-kit.jar" tofile="${basedir}/grammar-kit.jar"/>
+    </target>
+
+    <target name="psi.extract">
+        <copy file="${psiFile}" tofile="${basedir}/light-psi-all.jar"/>
+    </target>
+
+    <!--CHECK IF READY TO BUILD-->
+    <target name="ready">
+        <condition property="isReady">
+            <and>
```

Comment:
This is all that's necessary to build.  If we want to work with the lexers, though, we need the `idea-flex.skeleton` file, which comes from the IDEA community-edition sources: https://github.com/JetBrains/intellij-community/blob/master/tools/lexer/idea-flex.skeleton.  It would be nice to get that into place so that we can eventually remove the generated lexer files and only keep the .flex files in the source tree.

Reply:
I could probably add a step that fetches that file and add it to the workspace, I just need to know where to put it and if we are always going to get it from the master branch or if we should use a more "stable" branch where we won’t have to worry about major changes like package/folder refactoring etc.

Reply:
It goes into the same place that the grammar-kit jar goes: ${basedir}.

The best thing to do is pick it up from the tag matching the build we are making.  It's not terribly easy, though:  After unpacking the idea build, check the file `build.txt`.  It's in the root directory of the package on Linux and Windows.  The content of that file looks like `IU-172.3968.16`.   Dropping the leading portion of that, leaves us with the build number, which, when prepended with `idea/` gives us the git tag name for the build: `idea/172.3968.16`.  However, github doesn't allow us to reach a file by tag (just a source zip).  But, JetBrains always creates a branch for each release.  We can use that branch, which should be close enough, like so: 
`https://github.com/JetBrains/intellij-community/blob/172.3968/tools/lexer/idea-flex.skeleton`.

Reply:
Ok ill give it a shot, but i might need a few days :)

Reply:
Sure.  No worries.

Reply:
Just remembered that there is some code to deal with the build version in https://github.com/HaxeFoundation/intellij-haxe/blob/master/common.xml#L47-L148.  I'm not sure if you can re-use it, but it's there for reference, anyway.